pep-0723.rst

```diff
@@ -0,0 +1,540 @@
+=============
+
+Any Python script may assign a variable named ``__pyproject__`` to a multi-line
+*double-quoted* string (``"""``) containing a valid TOML document. The opening
+of the string MUST be on the same line as the assignment. The closing of the
+string MUST be on a line by itself, and MUST NOT be indented.
+
+The TOML document MUST NOT contain multi-line double-quoted strings, as that
+would conflict with the Python string containing the document. Single-quoted
+multi-line TOML strings may be used instead.
+
+This document MUST be formatted according to :pep:`621` and the
+`project metadata <pyproject metadata_>`_ standard. The document MAY include
+the ``[tool]`` table and sub-tables as described in :pep:`518`.
+
+The ``[project]`` table differs in the following ways:
+
+* The ``name`` and ``version`` fields are not required and MAY be defined
+  dynamically by tools if the user does not define them
+* These fields do not need to be listed in the ``dynamic`` array
+
+Non-script running tools MAY choose to read from their expected ``[tool]``
+sub-table if the script is the only target of the tool's functionality. In all
+other cases tools MUST NOT alter behavior based on the embedded metadata. For
+example, if a linter is invoked with the path to a directory, it MUST behave
+the same as if zero files had embedded metadata.
+
+Example
+-------
+
+The following is an example of a script with an embedded ``pyproject.toml``:
+
+.. code:: python
+
+    __pyproject__ = """
+    [project]
+    requires-python = ">=3.11"
+    dependencies = [
+      "requests<3",
+      "rich",
+    ]
+    """
+
+    import requests
+    from rich.pretty import pprint
+
+    resp = requests.get("https://peps.python.org/api/peps.json")
+    data = resp.json()
+    pprint([(k, v["title"]) for k, v in data.items()][:10])
+
+The following is an example of a single-file Rust project that embeds their
+version of ``pyproject.toml``, which is called ``Cargo.toml``:
+
+.. code:: rust
+
+    #!/usr/bin/env cargo
+
+    //! ```cargo
+    //! [dependencies]
+    //! regex = "1.8.0"
+    //! ```
+
+    fn main() {
+        let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
+        println!("Did our date match? {}", re.is_match("2014-01-01"));
+    }
+
+One important thing to note is that the metadata is embedded in a comment mostly
+for introspection since Rust documentation is generated from comments. Another
+is that users rarely edit dependencies manually, but rather use their Cargo
+package manager.
+
+We argue that our choice provides easier edits for both humans and tools.
+
+
+Reference Implementation
+========================
+
+This is the canonical regular expression that may be used to parse the metadata::
+
+    (?ms)^__pyproject__ *= *"""\\?$(.+?)^"""$
+
+For languages that do not support easily accessing the match group, one may
+parse the entire match as TOML (that is valid syntax) and then access
+the ``__pyproject__`` key from the resulting mapping.
```

Comment:
This still doesn't make sense -- the match is the text of a TOML table -- ``__pyproject__`` is the *Python* variable, not a sub-table within the TOML data.

Reply:
I can understand what it means. The *whole* of the text:

```
__pyproject__ = """
...
"""
```

can be parsed as TOML, giving a dictionary with a single key, `__pyproject__` and a value equal to the string value. If you then parse `data["__pyproject__"]` as TOML, you get the structure you want. (BTW, this is why I consider TOML to be a lot less "obvious" than people claim - this is *not* a straightforward TOML document...)

But (a) the description is almost incomprehensible, and (b) it's an *incredibly* convoluted way of getting at the data.

The text *definitely* needs to be revised, and it may be better to simply omit this option altogether, and simply assume that either the tool has to support extracting match groups from a RE match, or it'll have to work out how to parse the data for itself.

I don't know how common regex implementations are that can't extract the value a given match group. This may be a huge debate over nothing. Or it may be a major flaw in the proposal. It's hard to tell, TBH.

Reply:
I removed this bit as you are correct in saying there may be zero regular expression implementations that cannot support groups